### PR TITLE
Llama lora fully on device

### LINKED ISSUE #125: Move optimizer to device

### PR DESCRIPTION
**Pull request by pglusacTT**
Closes #125 #126  

### Problem description
Only forward pass was done on device for llama.

### What's changed
Moved everything to device.
Uplift ffe to 27.08.2025.

### CODE REVIEW

File: blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py [hunk 1]
```diff
@@ -9,47 +9,123 @@
 from tqdm import tqdm
 import wandb
 
+os.environ["WANDB_START_METHOD"] = "fork"
```
Comment (pmarkovicTT) [inline blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py:12]: curious, why this?

File: blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py [hunk 2]
```diff
@@ -59,7 +135,13 @@ def train(config, model, tokenizer, train_data_loader, val_data_loader):
...
         compiler_cfg = CompilerConfig()
         if config.dtype == "torch.bfloat16":
-            compiler_cfg.default_df_override = DataFormat.Float16
+            compiler_cfg.default_df_override = DataFormat.Float16_b
+            dtype = torch.bfloat16
+        elif config.dtype == "torch.float32":
+            dtype = torch.float32
+        else:
+            raise ValueError(f"Invalid dtype: {config.dtype}")
+        compiler_cfg.enable_consteval = False
```
Comment (pmarkovicTT) [inline blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py:144]: why do we need this?

File: blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py [hunk 3]
```diff
@@ -83,20 +165,39 @@ def train(config, model, tokenizer, train_data_loader, val_data_loader):
...
-    # Create a torch loss and leave on CPU
-    # Can be changed when https://github.com/tenstorrent/tt-metal/issues/18997 resolved
-    loss_fn = torch.nn.CrossEntropyLoss(ignore_index=-100)
+    vocab_size = model.model.config.vocab_size
+
+    if config.use_tt:
+        from blacksmith.experiments.torch.llama.loss import CrossEntropyLoss
+        loss_tt = CrossEntropyLoss(name="cross_entropy_loss")
+
+        seq_len_sample = config.max_length
```
Comment (pmarkovicTT) [inline blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py:183]: nit: since it is used only in one place (if I am not wrong), no need for new variable

File: blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py (hunk 2 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py]: Debugging purposes, now removed

File: blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py (hunk 3 above)
Comment (pglusacTT) [reply blacksmith/experiments/torch/llama/test_llama_fine_tuning_pure_torch.py]: Done in [78a7a71](https://github.com/tenstorrent/tt-blacksmith/pull/196/commits/78a7a71ff2970d6d804a3260313b8923c8cbfe70)

File: blacksmith/models/torch/huggingface/hf_models.py [hunk 4]
```diff
@@ -22,9 +22,11 @@ def forward(self, input_ids, attention_mask=None):
             causal_attention_mask = _prepare_4d_causal_attention_mask(
                 attention_mask, input_ids.shape, inputs_embeds, past_key_values_length
             )
-            logits = self.model(attention_mask=causal_attention_mask, inputs_embeds=inputs_embeds).logits
+            logits = self.model(attention_mask=causal_attention_mask, inputs_embeds=inputs_embeds).logits  # [B, T, V]
+            logits = logits.view(-1, logits.shape[-1]).t()  # [Batch * Seq len, Vocab size]
         else:
-            logits = self.model(input_ids=input_ids).logits
+            logits = self.model(input_ids=input_ids).logits  # [B, T, V]
+            logits = logits.view(-1, logits.shape[-1]).t()  # [Batch * Seq len, Vocab size]
```
[the remark this answers is not in the document]
Comment (pglusacTT) [reply blacksmith/models/torch/huggingface/hf_models.py]: Good catch for moving the reshape outside of if/else, done here [f8a5526](https://github.com/tenstorrent/tt-blacksmith/pull/196/commits/f8a55268760e5f3231a58c5c88cdb1486061cfea)
We are doing this when calling loss only for non-TT workloads, this needs to be done here for TT.
Non-TT doesn't use TextModelWrapper
